mesa/visualization/components/matplotlib.py

```diff
@@ -23,12 +25,42 @@ def SpaceMatplotlib(model, agent_portrayal, dependencies: list[any] | None = Non
     solara.FigureMatplotlib(space_fig, format="png", dependencies=dependencies)
 
 
+# matplotlib scatter does not allow for multiple shapes in one call
+def _split_and_scatter(portray_data, space_ax):
+    grouped_data = defaultdict(lambda: {"x": [], "y": [], "s": [], "c": []})
+
+    if "marker" not in portray_data:
+        # standard scatter is fine, default marker
+        space_ax.scatter(**portray_data)
+        return
+
+    # Extract data from the dictionary
+    markers = portray_data["marker"]
+    x = portray_data["x"]
+    y = portray_data["y"]
+    s = portray_data["s"]
+    c = portray_data["c"]
+
+    # Group the data by marker
+    for i in range(len(x)):
+        marker = markers[i]
+        grouped_data[marker]["x"].append(x[i])
+        grouped_data[marker]["y"].append(y[i])
+        grouped_data[marker]["s"].append(s[i])
+        grouped_data[marker]["c"].append(c[i])
+
+    # Plot each group with the same marker
+    for marker, data in grouped_data.items():
```

Comment:
I'm not sure of the performance implication of plotting each points separately. What about https://stackoverflow.com/a/43528356?

Reply:
Points are grouped by marker (shape) before plotting, like in the link provided.

Reply:
one way to potentially speed the grouping up is to use numpy.
if the x, y, s, c, markers arrays were np arrays, masks could be used on each np.unique() marker

Reply:
But I understand not wanting to introduce dependencies if not strictly necessary

Reply:
I want to add that NumPy has a lot of overhead. It is faster only for large arrays.